compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/InteractionUIView.uikit.kt

```diff
@@ -39,52 +50,354 @@ internal enum class CupertinoTouchesPhase {
+     */
+    override fun touchesCancelled(touches: Set<*>, withEvent: UIEvent?) {
+        stopTrackingTouches(touches)
+
+        onTouchesEvent(touches, withEvent, CupertinoTouchesPhase.CANCELLED)
+
+        if (hitTestView == view) {
+            // Golden path, just update the gesture recognizer state.
+
+            when (state) {
+                UIGestureRecognizerStateBegan, UIGestureRecognizerStateChanged -> {
+                    if (trackedTouches.isEmpty()) {
+                        state = UIGestureRecognizerStateCancelled
+                    } else {
+                        state = UIGestureRecognizerStateChanged
+                    }
+                }
+            }
+        } else {
+            if (trackedTouches.isEmpty()) {
+                // Those were the last touches in the sequence
+                // Explicitly fail the gesture, cancelling a scheduled failure
+                gestureRecognizer?.cancelFailure()
+
+                state = UIGestureRecognizerStateFailed
+            }
+        }
+    }
 
-        onTouchesEvent(view, touches, event, CupertinoTouchesPhase.ENDED)
+    /**
+     * Implementation of [CMPGestureRecognizerHandlerProtocol] that handles the failure of
+     * the gesture if it's not recognized within the certain time frame.
+     *
+     * It means we need to pass all the tracked touches to the runtime as cancelled and set failed
+     * state on the gesture recognizer.
+     *
+     * Intercepted touches will be passed to the interop views by UIKit due to
+     * [UIGestureRecognizer.delaysTouchesBegan]
+     */
+    override fun onFailure() {
+        state = UIGestureRecognizerStateFailed
+
+        // We won't receive other touches events until all fingers are lifted, so we can't rely
+        // on touchesEnded/touchesCancelled to reset the state.  We need to immediately notify
+        // the runtime about the cancelled touches and reset the state manually
+        onTouchesEvent(trackedTouches, null, CupertinoTouchesPhase.CANCELLED)
+        stopTrackingTouches(trackedTouches)
     }
 
-    override fun touchesCancelled(touches: Set<*>, withEvent: UIEvent?) {
-        onTouchesCountChanged(-touches.size)
+    /**
+     * Intentionally clean up all dependencies of GestureRecognizerHandlerImpl to prevent retain cycles that
+     * can be caused by implicit capture of the view by UIKit objects (such as UIEvent).
+     */
+    fun dispose() {
+        onTouchesEvent = { _, _, _, _ -> }
+        gestureRecognizer = null
+        trackedTouches.clear()
+    }
 
-        val view = view ?: return
-        val event = withEvent ?: return
+    /**
+     * Starts tracking the given touches. Remember initial location if those are the first touches
+     * in the sequence.
+     * @return `true` if the touches are initial, `false` otherwise.
+     */
+    private fun startTrackingTouches(touches: Set<*>): Boolean {
+        onTouchesCountChanged(touches.size)
```

Comment:
I would move this call on after we actually remove touches from trackedTouches

Reply:
Done